internal/pki/common.go

```diff
@@ -15,40 +13,94 @@ package pki
+	"crypto/x509/pkix"
 	"math/big"
 	"time"
 )
 
-const (
-	// beforeInterval sets a starting time for the issuance of a certificate,
-	// which is defaulted to an hour earlier
-	beforeInterval = -1 * time.Hour
+// certificateSignatureAlgorithm is ECDSA with SHA-384, the recommended
+// signature algorithm with the P-256 curve.
+const certificateSignatureAlgorithm = x509.ECDSAWithSHA384
 
-	// certificateSignatureAlgorithm sets the default signature algorithm to use
-	// for our certificates, which is the ECDSA with SHA-384. This is the
-	// recommended signature algorithm with the P-256 curve.
-	certificateSignatureAlgorithm = x509.ECDSAWithSHA384
+// currentTime returns the current local time. It is a variable so it can be
+// replaced during testing.
+var currentTime = time.Now
 
-	// serialNumberBits is the number of bits to allow in the generation of a
-	// random serial number
-	serialNumberBits = 128
-)
-
-// generateKey generates a ECDSA keypair using a P-256 curve. This curve is
-// roughly equivalent to a RSA 3072 bit key, but requires less bits to achieve
+// generateKey returns a random ECDSA key using a P-256 curve. This curve is
+// roughly equivalent to an RSA 3072-bit key but requires less bits to achieve
 // the equivalent cryptographic strength. Additionally, ECDSA is FIPS 140-2
 // compliant.
 func generateKey() (*ecdsa.PrivateKey, error) {
 	return ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
 }
 
-// generateSerialNumber generates a random serial number that can be used for
-// uniquely identifying a certificate
+// generateSerialNumber returns a random 128-bit integer.
 func generateSerialNumber() (*big.Int, error) {
-	return rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), serialNumberBits))
+	return rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+}
+
+func generateLeafCertificate(
+	signer *x509.Certificate, signerPrivate *ecdsa.PrivateKey,
+	signeePublic *ecdsa.PublicKey, serialNumber *big.Int,
+	commonName string, dnsNames []string,
+) (*x509.Certificate, error) {
+	const leafExpiration = time.Hour * 24 * 365
+	const leafStartValid = time.Hour * -1
+
+	now := currentTime()
+	template := &x509.Certificate{
+		BasicConstraintsValid: true,
+		DNSNames:              dnsNames,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		NotBefore:             now.Add(leafStartValid),
+		NotAfter:              now.Add(leafExpiration),
+		SerialNumber:          serialNumber,
+		SignatureAlgorithm:    certificateSignatureAlgorithm,
+		Subject: pkix.Name{
+			CommonName: commonName,
+		},
+	}
+
+	bytes, err := x509.CreateCertificate(rand.Reader, template, signer,
+		signeePublic, signerPrivate)
+
+	parsed, _ := x509.ParseCertificate(bytes)
+	return parsed, err
+}
+
+func generateRootCertificate(
+	privateKey *ecdsa.PrivateKey, serialNumber *big.Int,
+) (*x509.Certificate, error) {
+	const rootCommonName = "postgres-operator-ca"
+	const rootExpiration = time.Hour * 24 * 365 * 10
```

Comment:
Wait, sorry, just double-checking -- 10 years for the root and a year for the leaf?

Reply:
Yep, same as before.

https://github.com/CrunchyData/postgres-operator/blob/d0c3dae8236c7b7dd40f7003b21ebd9359871d92/internal/pki/root.go#L39-L41